contracts/ForwarderV4.sol

```diff
@@ -0,0 +1,353 @@
+import './ERC20Interface.sol';
+import './TransferHelper.sol';
+import './IForwarder.sol';
+
+/**
+ * @title ForwarderV4
+ * @notice This contract will forward any incoming Ether or token to the parent address of the contract
+ */
+contract ForwarderV4 is IERC721Receiver, ERC1155Receiver, IForwarder {
+  /// @notice Any funds sent to this contract will be forwarded to this address
+  address public parentAddress;
+  /// @notice Address which is allowed to call methods on this contract alongwith the parentAddress
+  address public feeAddress;
+  bool public autoFlush721 = true;
+  bool public autoFlush1155 = true;
+
+  /**
+   * @notice Event triggered when a deposit is received in the forwarder
+   * @param from Address from which the deposit is received
+   * @param value Amount of Ether received
+   * @param data Data sent along with the deposit
+   */
+  event ForwarderDeposited(address from, uint256 value, bytes data);
+
+  /**
+   * @notice Modifier that will execute internal code block only if the sender is from the allowed addresses
+   */
+  modifier onlyAllowedAddress() {
+    require(
+      msg.sender == parentAddress || msg.sender == feeAddress,
+      'Address is not allowed'
+    );
+    _;
+  }
+
+  /**
+   * @notice Modifier that will execute internal code block only if the contract has not been initialized yet
+   */
+  modifier onlyUninitialized() {
+    require(parentAddress == address(0x0), 'Already initialized');
+    _;
+  }
+
+  /**
+   * @notice Default function; Gets called when Ether is deposited with no data, and forwards it to the parent address
+   */
+  receive() external payable {
+    flush();
+  }
+
+  /**
+   * @notice Default function; Gets called when data is sent but does not match any other function
+   */
+  fallback() external payable {
+    flush();
+  }
+
+  /**
+   * @notice Initialize the contract, and sets the destination address to that of the parent address
+   * @param _parentAddress Address to which the funds should be forwarded
+   * @param _feeAddress Address which is allowed to call methods on this contract alongwith the parentAddress
+   * @param _autoFlush721 Whether to automatically flush ERC721 tokens or not
+   * @param _autoFlush1155 Whether to automatically flush ERC1155 tokens or not
+   */
+  function init(
+    address _parentAddress,
+    address _feeAddress,
+    bool _autoFlush721,
+    bool _autoFlush1155
+  ) external onlyUninitialized {
+    require(_parentAddress != address(0x0), 'Invalid address');
+    parentAddress = _parentAddress;
+    require(_feeAddress != address(0x0), 'Invalid address');
+    feeAddress = _feeAddress;
+
+    uint256 value = address(this).balance;
+
+    /// @notice set whether we want to automatically flush erc721/erc1155 tokens or not
+    autoFlush721 = _autoFlush721;
+    autoFlush1155 = _autoFlush1155;
+
+    if (value == 0) {
+      return;
+    }
+
+    (bool success, ) = parentAddress.call{ value: value }('');
```

Comment:
remind me why we use call instead of transfer. Transfer has protection against re-entry

Reply:
Any smart contract that uses transfer() is taking a hard dependency on gas costs by forwarding a fixed amount of gas: 2300. As gas costs are subject to change, then smart contracts can’t depend on any particular gas costs. Hence call method is the recommended method to transfer eth